src/libDirectoryService/DSBlockPreProcessing.cpp

```diff
@@ -209,27 +209,90 @@ void DirectoryService::ComputeSharding(const VectorOfPoWSoln& sortedPoWSolns) {
   }
 }
 
-void DirectoryService::InjectPoWForDSNode(VectorOfPoWSoln& sortedPoWSolns,
-                                          unsigned int numOfProposedDSMembers) {
-  // Add the oldest n DS committee member to m_allPoWs and m_allPoWConns so it
-  // gets included in sharding structure
-  if (numOfProposedDSMembers > m_mediator.m_DSCommittee->size()) {
+void DirectoryService::InjectPoWForDSNode(
+    VectorOfPoWSoln& sortedPoWSolns, std::map<PubKey, Peer>& powDSWinners) {
+  // Calculate the number of removed members from the candidates.
+  unsigned int numOfRemovedMembers = 0;
+  for (const auto& DSCandidate : powDSWinners) {
+    if (CheckIfDSNode(DSCandidate.first)) {
+      ++numOfRemovedMembers;
+    }
+  }
+
+  unsigned int numOfWinners = powDSWinners.size() - numOfRemovedMembers;
+  unsigned int numOfExpiring = numOfWinners - numOfRemovedMembers;
+
+  // Check the computed parameters for correctness.
+  if (powDSWinners.size() > m_mediator.m_DSCommittee->size()) {
     LOG_GENERAL(WARNING,
                 "FATAL: number of proposed ds member is larger than current ds "
                 "committee. numOfProposedDSMembers: "
-                    << numOfProposedDSMembers << " m_DSCommittee size: "
+                    << powDSWinners.size() << " m_DSCommittee size: "
                     << m_mediator.m_DSCommittee->size());
     return;
   }
 
+  if (numOfRemovedMembers > powDSWinners.size()) {
```

Comment:
Is this check necessary?

Reply:
Not really, but it helps with being paranoid and not making so many assumptions about the incoming data.

Reply:
Removed the unnecessary checks with the newly added field in the latest commit.